modAionImpl/src/org/aion/zero/impl/db/PendingBlockStore.java

```diff
@@ -0,0 +1,816 @@
+                if (!currentQ.isEmpty()) {
+                    // get first block in remaining queue
+                    AionBlock first = currentQ.get(0);
+
+                    // update queue hash to first remaining element
+                    byte[] currentQueueHash = first.getHash();
+
+                    // put in queue database
+                    queueSource.putToBatch(currentQueueHash, currentQ);
+
+                    // update block index
+                    for (AionBlock b : currentQ) {
+                        indexSource.putToBatch(b.getHash(), currentQueueHash);
+                    }
+
+                    // add (to) level
+                    byte[] levelKey = ByteUtil.longToBytes(first.getNumber());
+                    List<byte[]> levelData = levelSource.get(levelKey);
+
+                    if (levelData == null) {
+                        levelData = new ArrayList<>();
+                    }
+
+                    levelData.add(currentQueueHash);
+                    levelSource.putToBatch(levelKey, levelData);
+                }
+            }
+
+            // update level
+            byte[] levelKey = ByteUtil.longToBytes(level);
+            List<byte[]> levelData = levelSource.get(levelKey);
+
+            if (levelData == null) {
+                LOG.error(
+                        "Corrupt data in PendingBlockStorage. Level (expected to exist) was not found.");
+                // level already missing so nothing to do here
+            } else {
+                List<byte[]> updatedLevelData = new ArrayList<>();
+
+                levelData.forEach(
+                        qHash -> {
+                            if (!queues.contains(ByteArrayWrapper.wrap(qHash))) {
+                                // this queue was not imported
+                                updatedLevelData.add(qHash);
+                            }
+                        });
+
+                if (updatedLevelData.isEmpty()) {
+                    // delete level
+                    levelSource.putToBatch(levelKey, null);
+                } else {
+                    // update level
+                    levelSource.putToBatch(levelKey, updatedLevelData);
+                }
+            }
+
+            // push changed to disk
+            indexSource.commitBatch();
+            queueSource.flushBatch();
+            levelSource.flushBatch();
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    /**
+     * Generates a number greater or equal to the given {@code current} number representing the base
+     * value for a subsequent LIGHTNING request. The returned base is generated taking into
+     * consideration the status updates from {@link #addStatusBlock(AionBlock)} and the {@code
+     * knownBest} value for the peer for which this functionality is requested.
+     *
+     * <p>The bases are generated in an optimistic continuous manner based on the following
+     * assumptions:
```

Comment:
</p>

Reply:
The `<p>` is added by the google formatter. It does not require a closing tag: https://www.w3.org/TR/html401/intro/sgmltut.html#h-3.2.1